tools/make_patch.c

```diff
@@ -0,0 +1,437 @@
+		int padding = argc > 1 ? parse_number(argv[1], 0) : 2;
+		if (!strcmp(command, "HEx")) {
+			fprintf(output, "0x%0*X%02x", padding - 2, value >> 8, value & 0xff);
+		} else if (!strcmp(command, "Hex")) {
+			fprintf(output, "0x%0*X%03x", padding - 3, value >> 12, value & 0xfff);
+		} else if (!strcmp(command, "heX")) {
+			fprintf(output, "0x%0*x%02X", padding - 2, value >> 8, value & 0xff);
+		} else if (!strcmp(command, "hEX")) {
+			fprintf(output, "0x%0*x%03X", padding - 3, value >> 12, value & 0xfff);
+		} else {
+			fprintf(output, isupper((unsigned char)command[0]) ? "0x%0*X" : "0x%0*x", padding, value);
+		}
+
+	} else {
+		error_exit("Error: Unknown command: \"%s\"\n", command);
+	}
+}
+
+void skip_to_next_line(FILE *restrict input, FILE *restrict output) {
+	for (int c = getc(input); c != EOF; c = getc(input)) {
+		putc(c, output);
+		if (c == '\n' || c == '\r') {
+			break;
+		}
+	}
+}
+
+struct Buffer *process_template(const char *template_filename, const char *patch_filename, FILE *restrict new_rom, FILE *restrict orig_rom, const struct Symbol *symbols) {
+	FILE *input = xfopen(template_filename, 'r');
+	FILE *output = xfopen(patch_filename, 'w');
+	struct Buffer *buffer = buffer_create(1);
+
+	struct Buffer *patches = buffer_create(sizeof(struct Patch));
+
+	// The ROM checksum will always differ
+	buffer_append(patches, &(struct Patch){0x14e, 2});
+	// The Stadium data (see stadium.c) will always differ
+	unsigned int rom_size = (unsigned int)xfsize("", orig_rom);
+	unsigned int stadium_size = 24 + 6 + 2 + (rom_size / 0x2000) * 2;
+	buffer_append(patches, &(struct Patch){rom_size - stadium_size, stadium_size});
+
+	const struct Symbol *current_hook = NULL;
+
+	// Fill in the template
+	for (int c = getc(input); c != EOF; c = getc(input)) {
+		switch (c) {
+		case ';':
+			// ";" comments until the end of the line
+			putc(c, output);
+			skip_to_next_line(input, output);
+			break;
+
+		case '{':
+			// "{...}" is a template command; buffer its contents
+			buffer->size = 0;
+			for (c = getc(input); c != EOF && c != '}'; c = getc(input)) {
+				buffer_append(buffer, &c);
+			}
+			buffer_append(buffer, &(char []){'\0'});
+			// Interpret the command in the context of the current patch
+			interpret_command(buffer->data, current_hook, symbols, patches, new_rom, orig_rom, output);
+			break;
+
+		case '[':
+			// "[...]" is a patch label; buffer its contents
+			putc(c, output);
+			buffer->size = 0;
+			for (c = getc(input); c != EOF; c = getc(input)) {
+				putc(c, output);
+				if (c == ']') {
+					break;
+				} else if (!isalnum(c) && c != '_' && c != '@' && c != '#') {
+					// Convert non-identifier characters to underscores
+					c = '_';
+				}
+				buffer_append(buffer, &c);
+			}
+			buffer_append(buffer, &(char []){'\0'});
+			// The current patch should have a corresponding ".VC_" label
+			current_hook = symbol_find_cat(symbols, ".VC_", buffer->data);
```

Comment:
This implementation is still lacking a method for setting a different name for the current hook (one that doesn't match the original name). I'd propose a syntax like `[Network363:literally_any_other_name_would_be_more_descriptive]`, where the `:` and anything after it isn't output.

Reply:
Do we ever need such a thing? If so I'd recommend `[Network363@a_more_sensible_name]` to match the `@` symbol.

Reply:
I can seem _some_ benefit to this. Although we can always implement it later when we know more about these more obscure hooks.

Reply:
Yeah, IMO it'll be worth adding if/when we actually have more specific+descriptive labels. Same goes for comments, if we want to add them in the template instead of in the asm.

Reply:
I can already think of a few use cases for this. We already know what some of the different print_forbids do. I'd like to have print_forbid_box, print_forbid_dex, print_forbid_seer, etc.

Reply:
Oh and `FPA 006 Begin` becomes `flash_dim_horn_drill`, do the same for the rest of the FPAs.